File: backend/app/api/v1/endpoints/management.py
from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Query

from app.core.dependencies import AdminOrSubAdminOrFinancial, CurrentUserPayload, DBSession
from app.domain.schemas.common import PaginatedResponse
from app.domain.schemas.property_management import (
    ManagementEntityCreate,
    ManagementEntityResponse,
    ManagementEntityUpdate,
    OwnerCreate,
    OwnerResponse,
    OwnerUpdate,
    PropertyGroupCreate,
    PropertyGroupResponse,
    PropertyGroupUpdate,
    UnitManagementContractCreate,
    UnitManagementContractResponse,
    UnitManagementContractUpdate,
)
from app.services.classification_service import ClassificationService

router = APIRouter(prefix="/management", tags=["Management"])


@router.post("/owners", response_model=OwnerResponse, dependencies=[AdminOrSubAdminOrFinancial])
async def create_owner(data: OwnerCreate, db: DBSession):
    service = ClassificationService(db)
    return await service.create_owner(data)


@router.get("/owners", response_model=PaginatedResponse[OwnerResponse])
async def list_owners(
    db: DBSession,
    _: CurrentUserPayload,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    service = ClassificationService(db)
    items, total = await service.list_owners(skip=(page - 1) * page_size, limit=page_size)
    return PaginatedResponse.create(
        items=[OwnerResponse.model_validate(item) for item in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.patch("/owners/{owner_id}", response_model=OwnerResponse, dependencies=[AdminOrSubAdminOrFinancial])
async def update_owner(owner_id: UUID, data: OwnerUpdate, db: DBSession):
    service = ClassificationService(db)
    return await service.update_owner(owner_id, data)


@router.post("/entities", response_model=ManagementEntityResponse, dependencies=[AdminOrSubAdminOrFinancial])
async def create_management_entity(data: ManagementEntityCreate, db: DBSession):
    service = ClassificationService(db)
    return await service.create_management_entity(data)


@router.get("/entities", response_model=PaginatedResponse[ManagementEntityResponse])
async def list_management_entities(
    db: DBSession,
    _: CurrentUserPayload,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    service = ClassificationService(db)
    items, total = await service.list_management_entities(skip=(page - 1) * page_size, limit=page_size)
    return PaginatedResponse.create(
        items=[ManagementEntityResponse.model_validate(item) for item in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.patch("/entities/{entity_id}", response_model=ManagementEntityResponse, dependencies=[AdminOrSubAdminOrFinancial])
async def update_management_entity(entity_id: UUID, data: ManagementEntityUpdate, db: DBSession):
    service = ClassificationService(db)
    return await service.update_management_entity(entity_id, data)


@router.post("/property-groups", response_model=PropertyGroupResponse, dependencies=[AdminOrSubAdminOrFinancial])
async def create_property_group(data: PropertyGroupCreate, db: DBSession):
    service = ClassificationService(db)
    return await service.create_property_group(data)


@router.get("/property-groups", response_model=PaginatedResponse[PropertyGroupResponse])
async def list_property_groups(
    db: DBSession,
    _: CurrentUserPayload,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
):
    service = ClassificationService(db)
    items, total = await service.list_property_groups(skip=(page - 1) * page_size, limit=page_size)
    return PaginatedResponse.create(
        items=[PropertyGroupResponse.model_validate(item) for item in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.patch("/property-groups/{group_id}", response_model=PropertyGroupResponse, dependencies=[AdminOrSubAdminOrFinancial])
async def update_property_group(group_id: UUID, data: PropertyGroupUpdate, db: DBSession):
    service = ClassificationService(db)
    return await service.update_property_group(group_id, data)


@router.post("/contracts", response_model=UnitManagementContractResponse, dependencies=[AdminOrSubAdminOrFinancial])
async def create_contract(data: UnitManagementContractCreate, db: DBSession):
    service = ClassificationService(db)
    return await service.create_contract(data)


@router.get("/contracts", response_model=PaginatedResponse[UnitManagementContractResponse])
async def list_contracts(
    db: DBSession,
    _: CurrentUserPayload,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    unit_id: Optional[UUID] = Query(None),
):
    service = ClassificationService(db)
    items, total = await service.list_contracts(skip=(page - 1) * page_size, limit=page_size, unit_id=unit_id)
    return PaginatedResponse.create(
        items=[UnitManagementContractResponse.model_validate(item) for item in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.patch("/contracts/{contract_id}", response_model=UnitManagementContractResponse, dependencies=[AdminOrSubAdminOrFinancial])
async def update_contract(contract_id: UUID, data: UnitManagementContractUpdate, db: DBSession):
    service = ClassificationService(db)
    return await service.update_contract(contract_id, data)

File: frontend/lib/api/management.ts
import { apiClient } from "@/lib/api/client";
import type {
  ManagementEntity,
  Owner,
  PaginatedResponse,
  PropertyGroup,
  UnitManagementContract,
} from "@/types";

function toSearchParams(params?: object) {
  const searchParams = new URLSearchParams();
  if (params) {
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== null && value !== "") {
        searchParams.set(key, String(value));
      }
    });
  }
  return searchParams.toString();
}

export const managementApi = {
  listOwners: async (params?: { page?: number; page_size?: number }): Promise<PaginatedResponse<Owner>> => {
    const response = await apiClient.get<PaginatedResponse<Owner>>(`/management/owners?${toSearchParams(params)}`);
    return response.data;
  },
  createOwner: async (data: Partial<Owner>): Promise<Owner> => {
    const response = await apiClient.post<Owner>("/management/owners", data);
    return response.data;
  },
  updateOwner: async (id: string, data: Partial<Owner>): Promise<Owner> => {
    const response = await apiClient.patch<Owner>(`/management/owners/${id}`, data);
    return response.data;
  },

  listEntities: async (params?: { page?: number; page_size?: number }): Promise<PaginatedResponse<ManagementEntity>> => {
    const response = await apiClient.get<PaginatedResponse<ManagementEntity>>(`/management/entities?${toSearchParams(params)}`);
    return response.data;
  },
  createEntity: async (data: Partial<ManagementEntity>): Promise<ManagementEntity> => {
    const response = await apiClient.post<ManagementEntity>("/management/entities", data);
    return response.data;
  },
  updateEntity: async (id: string, data: Partial<ManagementEntity>): Promise<ManagementEntity> => {
    const response = await apiClient.patch<ManagementEntity>(`/management/entities/${id}`, data);
    return response.data;
  },

  listPropertyGroups: async (params?: { page?: number; page_size?: number }): Promise<PaginatedResponse<PropertyGroup>> => {
    const response = await apiClient.get<PaginatedResponse<PropertyGroup>>(`/management/property-groups?${toSearchParams(params)}`);
    return response.data;
  },
  createPropertyGroup: async (data: Partial<PropertyGroup>): Promise<PropertyGroup> => {
    const response = await apiClient.post<PropertyGroup>("/management/property-groups", data);
    return response.data;
  },
  updatePropertyGroup: async (id: string, data: Partial<PropertyGroup>): Promise<PropertyGroup> => {
    const response = await apiClient.patch<PropertyGroup>(`/management/property-groups/${id}`, data);
    return response.data;
  },

  listContracts: async (params?: { page?: number; page_size?: number; unit_id?: string }): Promise<PaginatedResponse<UnitManagementContract>> => {
    const response = await apiClient.get<PaginatedResponse<UnitManagementContract>>(`/management/contracts?${toSearchParams(params)}`);
    return response.data;
  },
  createContract: async (data: Partial<UnitManagementContract>): Promise<UnitManagementContract> => {
    const response = await apiClient.post<UnitManagementContract>("/management/contracts", data);
    return response.data;
  },
  updateContract: async (id: string, data: Partial<UnitManagementContract>): Promise<UnitManagementContract> => {
    const response = await apiClient.patch<UnitManagementContract>(`/management/contracts/${id}`, data);
    return response.data;
  },
};